test/fmf/tests/utils.sh

```diff
@@ -0,0 +1,406 @@
+log_warn() {
+    echo -e "${YELLOW}[WARN]${NC} 🚧 $1"
+}
+
+log_success() {
+    echo -e "${GREEN}[SUCCESS]${NC} ✔ $1"
+}
+
+log_debug() {
+    if [ "${DEBUG:-0}" = "1" ]; then
+        echo -e "${BLUE}[DEBUG]${NC} $1"
+    fi
+}
+
+test_pass() {
+    echo -e "${GREEN}[PASS]${NC} ✅ Test PASSED!"
+}
+
+test_fail() {
+    echo -e "${RED}[FAIL]${NC} ❌ Test FAILED!"
+    return 1
+}
+
+#==============================================================================
+# ENVIRONMENT SETUP
+#==============================================================================
+
+set_hostname() {
+    local dns=$1
+    local ip=$2
+    if grep -q " ${dns}" /etc/hosts; then
+        tmp_hosts=$(mktemp)
+        sed "s/.* ${dns}/${ip} ${dns}/" /etc/hosts >"${tmp_hosts}"
+        cp "${tmp_hosts}" /etc/hosts
+        rm -f "${tmp_hosts}"
+    else
+        echo "${ip} ${dns}" | sudo tee -a /etc/hosts > /dev/null
+    fi
+}
+
+setup_test_hostnames() {
+    log_info "Setting up test hostnames..."
+    set_hostname manufacturer 127.0.0.1
+    set_hostname rendezvous 127.0.0.1
+    set_hostname owner 127.0.0.1
+}
+
+verify_fdo_packages() {
+    log_info "Verifying FDO packages..."
+
+    if ! rpm -q go-fdo-client &>/dev/null; then
+        log_error "go-fdo-client package not installed"
+        return 1
+    fi
+
+    if ! rpm -q go-fdo-server-manufacturer go-fdo-server-rendezvous go-fdo-server-owner &>/dev/null; then
+        log_error "go-fdo-server packages not installed"
+        return 1
+    fi
+
+    if ! command -v go-fdo-client &> /dev/null; then
+        log_error "go-fdo-client binary not found in PATH"
+        return 1
+    fi
+
+    log_info "All FDO packages verified"
+    return 0
+}
+
+start_fdo_services() {
+    log_info "Starting FDO services..."
+
+    # Check if services are already running via port check
+    local ports_map="manufacturer:8038 rendezvous:8041 owner:8043"
+    local all_running=true
+
+    for entry in $ports_map; do
+        local service="${entry%:*}"
+        local port="${entry#*:}"
+        if ! ss -tln | grep -q ":${port} "; then
+            all_running=false
+            break
+        fi
+    done
+
+    if $all_running; then
+        log_info "FDO services already running (ports 8038, 8041, 8043 in use)"
+        return 0
+    fi
+
+    # Fix certificate permissions
+    chown go-fdo-server-manufacturer:go-fdo-server /etc/go-fdo-server/{manufacturer,device-ca}*.{crt,key} 2>/dev/null || true
+    chmod g+r /etc/go-fdo-server/device-ca.crt 2>/dev/null || true
+    chown go-fdo-server-owner:go-fdo-server /etc/go-fdo-server/owner*.{crt,key} 2>/dev/null || true
+    chmod g+r /etc/go-fdo-server/owner.crt 2>/dev/null || true
+
+    for service in manufacturer rendezvous owner; do
+        systemctl start go-fdo-server-${service}.service || {
+            log_error "Failed to start ${service} server"
+            journalctl -u go-fdo-server-${service}.service --no-pager
+            return 1
+        }
+    done
+
+    sleep 5
```

Comment:
You might be better off polling the server's health check url rather than sleeping and checking active.

Just copy and pasta the impl from [go-fdo-server](https://github.com/fido-device-onboard/go-fdo-server/blob/d37e23dca1ea2d601d35911973a8c7f0f3ade513/test/ci/utils.sh#L228)